From 328ff0919ff07284a6a0c881c935868a88505814 Mon Sep 17 00:00:00 2001
From: Sebastian Schaper <openwrt@sebastianschaper.net>
Date: Thu, 18 Mar 2021 19:33:04 +0100
Subject: [PATCH] ath79: add experimental support for ZyXEL NWA1121-NI /
 NWA1123-NI

Installation:
* OEM Web Interface is at 192.168.1.2, login with
   admin
   1234
* flash factory-XXXX.bin

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
---
 ...h79-add-support-for-ZyXEL-NWA1121-NI.patch | 289 ++++++++++++++++++
 ...h79-add-support-for-ZyXEL-NWA1123-NI.patch | 273 +++++++++++++++++
 targets/ath79-generic                         |   6 +
 3 files changed, 568 insertions(+)
 create mode 100644 patches/openwrt/0025-ath79-add-support-for-ZyXEL-NWA1121-NI.patch
 create mode 100644 patches/openwrt/0026-ath79-add-support-for-ZyXEL-NWA1123-NI.patch

diff --git a/patches/openwrt/0025-ath79-add-support-for-ZyXEL-NWA1121-NI.patch b/patches/openwrt/0025-ath79-add-support-for-ZyXEL-NWA1121-NI.patch
new file mode 100644
index 00000000..e0743dcb
--- /dev/null
+++ b/patches/openwrt/0025-ath79-add-support-for-ZyXEL-NWA1121-NI.patch
@@ -0,0 +1,289 @@
+From a72441ee5194b3319c4d83fecee6a4366c1ec65e Mon Sep 17 00:00:00 2001
+From: Sebastian Schaper <openwrt@sebastianschaper.net>
+Date: Wed, 9 Sep 2020 20:42:53 +0200
+Subject: [PATCH 1/2] ath79: add support for ZyXEL NWA1121-NI
+
+Specifications:
+ * AR9344, 8 MiB Flash, 64 MiB RAM
+ * 1x Gigabit Ethernet, 802.3af PoE
+
+Installation:
+* OEM Web UI is at 192.168.1.2
+  login as `admin` with password `1234`
+* flash factory-AABJ.bin
+
+TFTP Recovery:
+* open the case, solder serial port
+  (this is the official method described by Zyxel;
+  the reset button is useless during power-on)
+* extract factory image (.tar.bz2), provide
+  `vmlinux_mi124_f1e.lzma.uImage` and
+  `mi124_f1e-jffs2` via tftp at 192.168.1.10
+* interrupt uboot countdown, execute commands
+  `run lk`
+  `run lf`
+  to flash the kernel / filesystem accordingly
+
+Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
+---
+ .../ath79/dts/ar9342_zyxel_nwa1121-ni.dts     | 171 ++++++++++++++++++
+ .../generic/base-files/etc/board.d/02_network |   3 +-
+ .../etc/hotplug.d/ieee80211/10_fix_wifi_mac   |   3 +
+ target/linux/ath79/image/generic.mk           |  30 +++
+ 4 files changed, 206 insertions(+), 1 deletion(-)
+ create mode 100644 target/linux/ath79/dts/ar9342_zyxel_nwa1121-ni.dts
+
+diff --git a/target/linux/ath79/dts/ar9342_zyxel_nwa1121-ni.dts b/target/linux/ath79/dts/ar9342_zyxel_nwa1121-ni.dts
+new file mode 100644
+index 0000000000..271d1989b1
+--- /dev/null
++++ b/target/linux/ath79/dts/ar9342_zyxel_nwa1121-ni.dts
+@@ -0,0 +1,171 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++
++#include "ar9344.dtsi"
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++#include <dt-bindings/mtd/partitions/uimage.h>
++
++/ {
++	compatible = "zyxel,nwa1121-ni", "qca,ar9342";
++	model = "Zyxel NWA1121-NI";
++
++	aliases {
++		label-mac-device = &eth0;
++
++		led-boot = &led_status_green;
++		led-failsafe = &led_status_amber;
++		led-running = &led_status_green;
++		led-upgrade = &led_status_amber;
++	};
++
++	leds {
++		compatible = "gpio-leds";
++
++		led_status_green: status_green {
++			label = "green:status";
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++			default-state = "on";
++		};
++
++		led_status_amber: status_amber {
++			label = "amber:status";
++			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	keys {
++		compatible = "gpio-keys";
++
++		reset {
++			label = "reset";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	virtual_flash {
++		compatible = "mtd-concat";
++		devices = <&firmware0 &firmware1>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			partition@0 {
++				compatible = "openwrt,uimage", "denx,uimage";
++				openwrt,ih-magic = <IH_MAGIC_OKLI>;
++				label = "firmware";
++				reg = <0x0 0x0>;
++			};
++		};
++	};
++};
++
++&ref {
++	clock-frequency = <40000000>;
++};
++
++&spi {
++	status = "okay";
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <50000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x040000>;
++				read-only;
++			};
++
++			partition@40000 {
++				label = "u-boot-env";
++				reg = <0x040000 0x010000>;
++			};
++
++			firmware0: partition@50000 {
++				label = "firmware0";
++				reg = <0x50000 0x800000>;
++			};
++
++			partition@850000 {
++				label = "loader";
++				reg = <0x850000 0x010000>;
++				read-only;
++			};
++
++			firmware1: partition@860000 {
++				label = "firmware1";
++				reg = <0x860000 0x740000>;
++			};
++
++			partition@fa0000 {
++				label = "config";
++				reg = <0xfa0000 0x040000>;
++				read-only;
++			};
++
++			partition@fe0000 {
++				label = "mib0";
++				reg = <0xfe0000 0x10000>;
++				read-only;
++			};
++
++			art: partition@ff0000 {
++				label = "art";
++				reg = <0xff0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&mdio0 {
++	status = "okay";
++
++	phy-mask = <0>;
++
++	phy0: ethernet-phy@0 {
++		reg = <0>;
++		phy-mode = "rgmii";
++	};
++};
++
++&eth0 {
++	status = "okay";
++
++	pll-data = <0x06000000 0x00000101 0x00001313>;
++
++	mtd-mac-address = <&art 0x1002>;
++
++	phy-mode = "rgmii-id";
++	phy-handle = <&phy0>;
++
++	gmac-config {
++		device = <&gmac>;
++		rxdv-delay = <3>;
++		rxd-delay = <3>;
++		txen-delay = <3>;
++		txd-delay = <3>;
++		rgmii-gmac0 = <1>;
++	};
++};
++
++&uart {
++	status = "okay";
++};
++
++&wmac {
++	status = "okay";
++
++	qca,disable-5ghz;
++	mtd-cal-data = <&art 0x1000>;
++};
+diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
+index 86be2b12d7..fc99379682 100644
+--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
++++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
+@@ -92,7 +92,8 @@ ath79_setup_interfaces()
+ 	ubnt,unifiac-mesh|\
+ 	ubnt,unifi|\
+ 	wd,mynet-wifi-rangeextender|\
+-	winchannel,wb2000)
++	winchannel,wb2000|\
++	zyxel,nwa1121-ni)
+ 		ucidef_set_interface_lan "eth0"
+ 		;;
+ 	airtight,c-75)
+diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
+index ac8b59c538..0d8a15a780 100644
+--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
++++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
+@@ -65,4 +65,7 @@ case "$board" in
+ 		[ "$PHYNBR" -eq 1 ] && \
+ 			mtd_get_mac_ascii u-boot-env ethaddr > /sys${DEVPATH}/macaddress
+ 		;;
++	zyxel,nwa1121-ni)
++		mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
++		;;
+ esac
+diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
+index a746bcfb8e..9177ec9e8d 100644
+--- a/target/linux/ath79/image/generic.mk
++++ b/target/linux/ath79/image/generic.mk
+@@ -151,6 +151,14 @@ define Build/wrgg-pad-rootfs
+ 	$(STAGING_DIR_HOST)/bin/padjffs2 $(IMAGE_ROOTFS) -c 64 >>$@
+ endef
+ 
++define Build/zyxel-tar-bz2
++	mv $@ $(word 3,$(1))
++	mv $(BUILD_DIR)/linux-ath79_generic/loader-$(word 1,$(1)).uImage \
++		$(word 2,$(1)).lzma.uImage
++	tar --transform 's?.*/??g' -cjf $@ $(word 3,$(1)) \
++		$(word 2,$(1)).lzma.uImage
++	rm $(word 3,$(1)) $(word 2,$(1)).lzma.uImage
++endef
+ 
+ define Device/seama
+   KERNEL := kernel-bin | append-dtb | relocate-kernel | lzma
+@@ -2206,6 +2214,28 @@ define Device/zbtlink_zbt-wd323
+ endef
+ TARGET_DEVICES += zbtlink_zbt-wd323
+ 
++define Device/zyxel_nwa1121-ni
++  SOC := ar9342
++  DEVICE_VENDOR := Zyxel
++  DEVICE_MODEL := NWA1121
++  DEVICE_VARIANT := NI
++  COMPILE := loader-$(1).bin loader-$(1).uImage
++  COMPILE/loader-$(1).bin := loader-okli-compile
++  COMPILE/loader-$(1).uImage := append-loader-okli $(1) | pad-to 64k | \
++	lzma | uImage lzma
++  LOADER_FLASH_OFFS := 0x050000
++  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x4f4b4c49
++# todo: allow larger factory images (split + merge with uImage partition)
++  IMAGE_SIZE := 8192k
++  IMAGES += loader.bin sysupgrade.bin factory-AABJ.bin
++  IMAGE/loader.bin := append-loader-okli $(1) | pad-to 64k | lzma | \
++	uImage lzma
++  IMAGE/factory-AABJ.bin := $$(IMAGE/sysupgrade.bin) | pad-to 8192k | \
++	zyxel-tar-bz2 zyxel_nwa1121-ni vmlinux_mi124_f1e mi124_f1e-jffs2 | \
++	append-md5sum-bin
++endef
++TARGET_DEVICES += zyxel_nwa1121-ni
++
+ define Device/zyxel_nbg6616
+   SOC := qca9557
+   DEVICE_VENDOR := ZyXEL
+-- 
+2.25.1
+
diff --git a/patches/openwrt/0026-ath79-add-support-for-ZyXEL-NWA1123-NI.patch b/patches/openwrt/0026-ath79-add-support-for-ZyXEL-NWA1123-NI.patch
new file mode 100644
index 00000000..d33043f3
--- /dev/null
+++ b/patches/openwrt/0026-ath79-add-support-for-ZyXEL-NWA1123-NI.patch
@@ -0,0 +1,273 @@
+From 7ff2526dc6c6cd5f2b077f65032c033a07432e8d Mon Sep 17 00:00:00 2001
+From: Sebastian Schaper <openwrt@sebastianschaper.net>
+Date: Tue, 16 Mar 2021 00:01:50 +0100
+Subject: [PATCH 2/2] ath79: add support for ZyXEL NWA1123-NI
+
+Specifications:
+* identical to NWA1121-NI, with additional 5G ath9k radio
+
+Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
+---
+ .../ath79/dts/ar9342_zyxel_nwa1123-ni.dts     | 182 ++++++++++++++++++
+ .../generic/base-files/etc/board.d/02_network |   3 +-
+ .../etc/hotplug.d/ieee80211/10_fix_wifi_mac   |   8 +-
+ target/linux/ath79/image/generic.mk           |  22 +++
+ 4 files changed, 212 insertions(+), 3 deletions(-)
+ create mode 100644 target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts
+
+diff --git a/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts b/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts
+new file mode 100644
+index 0000000000..979c604971
+--- /dev/null
++++ b/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts
+@@ -0,0 +1,182 @@
++// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
++
++#include "ar9344.dtsi"
++
++#include <dt-bindings/gpio/gpio.h>
++#include <dt-bindings/input/input.h>
++#include <dt-bindings/mtd/partitions/uimage.h>
++
++/ {
++	compatible = "zyxel,nwa1123-ni", "qca,ar9342";
++	model = "Zyxel NWA1123-NI";
++
++	aliases {
++		label-mac-device = &eth0;
++
++		led-boot = &led_status_green;
++		led-failsafe = &led_status_amber;
++		led-running = &led_status_green;
++		led-upgrade = &led_status_amber;
++	};
++
++	leds {
++		compatible = "gpio-leds";
++
++		led_status_green: status_green {
++			label = "green:status";
++			gpios = <&gpio 12 GPIO_ACTIVE_LOW>;
++			default-state = "on";
++		};
++
++		led_status_amber: status_amber {
++			label = "amber:status";
++			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	keys {
++		compatible = "gpio-keys";
++
++		reset {
++			label = "reset";
++			linux,code = <KEY_RESTART>;
++			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
++		};
++	};
++
++	virtual_flash {
++		compatible = "mtd-concat";
++		devices = <&firmware0 &firmware1>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			partition@0 {
++				compatible = "openwrt,uimage", "denx,uimage";
++				openwrt,ih-magic = <IH_MAGIC_OKLI>;
++				label = "firmware";
++				reg = <0x0 0x0>;
++			};
++		};
++	};
++};
++
++&ref {
++	clock-frequency = <40000000>;
++};
++
++&spi {
++	status = "okay";
++
++	flash@0 {
++		compatible = "jedec,spi-nor";
++		reg = <0>;
++		spi-max-frequency = <50000000>;
++
++		partitions {
++			compatible = "fixed-partitions";
++			#address-cells = <1>;
++			#size-cells = <1>;
++
++			partition@0 {
++				label = "u-boot";
++				reg = <0x000000 0x040000>;
++				read-only;
++			};
++
++			partition@40000 {
++				label = "u-boot-env";
++				reg = <0x040000 0x010000>;
++			};
++
++			firmware0: partition@50000 {
++				label = "firmware0";
++				reg = <0x50000 0x800000>;
++			};
++
++			partition@850000 {
++				label = "loader";
++				reg = <0x850000 0x010000>;
++				read-only;
++			};
++
++			firmware1: partition@860000 {
++				label = "firmware1";
++				reg = <0x860000 0x740000>;
++			};
++
++			partition@fa0000 {
++				label = "config";
++				reg = <0xfa0000 0x040000>;
++				read-only;
++			};
++
++			partition@fe0000 {
++				label = "mib0";
++				reg = <0xfe0000 0x10000>;
++				read-only;
++			};
++
++			art: partition@ff0000 {
++				label = "art";
++				reg = <0xff0000 0x010000>;
++				read-only;
++			};
++		};
++	};
++};
++
++&mdio0 {
++	status = "okay";
++
++	phy-mask = <0>;
++
++	phy0: ethernet-phy@0 {
++		reg = <0>;
++		phy-mode = "rgmii";
++	};
++};
++
++&eth0 {
++	status = "okay";
++
++	pll-data = <0x06000000 0x00000101 0x00001313>;
++
++	mtd-mac-address = <&art 0x1002>;
++
++	phy-mode = "rgmii-id";
++	phy-handle = <&phy0>;
++
++	gmac-config {
++		device = <&gmac>;
++		rxdv-delay = <3>;
++		rxd-delay = <3>;
++		txen-delay = <3>;
++		txd-delay = <3>;
++		rgmii-gmac0 = <1>;
++	};
++};
++
++&pcie {
++	status = "okay";
++
++	ath9k: wifi@0,0,0 {
++		compatible = "pci168c,0030";
++		reg = <0x0 0 0 0 0>;
++		#mtd-mac-address = <&art 0x1002>;
++		#mtd-mac-address-increment = <3>;
++	};
++};
++
++&uart {
++	status = "okay";
++};
++
++&wmac {
++	status = "okay";
++
++	qca,disable-5ghz;
++	mtd-cal-data = <&art 0x1000>;
++};
+diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
+index fc99379682..5c8e2f15fc 100644
+--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
++++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
+@@ -93,7 +93,8 @@ ath79_setup_interfaces()
+ 	ubnt,unifi|\
+ 	wd,mynet-wifi-rangeextender|\
+ 	winchannel,wb2000|\
+-	zyxel,nwa1121-ni)
++	zyxel,nwa1121-ni|\
++	zyxel,nwa1123-ni)
+ 		ucidef_set_interface_lan "eth0"
+ 		;;
+ 	airtight,c-75)
+diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
+index 0d8a15a780..73844df275 100644
+--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
++++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
+@@ -65,7 +65,11 @@ case "$board" in
+ 		[ "$PHYNBR" -eq 1 ] && \
+ 			mtd_get_mac_ascii u-boot-env ethaddr > /sys${DEVPATH}/macaddress
+ 		;;
+-	zyxel,nwa1121-ni)
+-		mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
++	zyxel,nwa1121-ni|\
++	zyxel,nwa1123-ni)
++		[ "$PHYNBR" -eq 0 ] && \
++			mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
++		[ "$PHYNBR" -eq 1 ] && \
++			mtd_get_mac_text mib0 0x66 > /sys${DEVPATH}/macaddress
+ 		;;
+ esac
+diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
+index 9177ec9e8d..0e85382661 100644
+--- a/target/linux/ath79/image/generic.mk
++++ b/target/linux/ath79/image/generic.mk
+@@ -2236,6 +2236,28 @@ define Device/zyxel_nwa1121-ni
+ endef
+ TARGET_DEVICES += zyxel_nwa1121-ni
+ 
++define Device/zyxel_nwa1123-ni
++  SOC := ar9342
++  DEVICE_VENDOR := Zyxel
++  DEVICE_MODEL := NWA1123
++  DEVICE_VARIANT := NI
++  COMPILE := loader-$(1).bin loader-$(1).uImage
++  COMPILE/loader-$(1).bin := loader-okli-compile
++  COMPILE/loader-$(1).uImage := append-loader-okli $(1) | pad-to 64k | \
++	lzma | uImage lzma
++  LOADER_FLASH_OFFS := 0x050000
++  KERNEL := kernel-bin | append-dtb | lzma | uImage lzma -M 0x4f4b4c49
++# todo: allow larger factory images (split + merge with uImage partition)
++  IMAGE_SIZE := 8192k
++  IMAGES += loader.bin sysupgrade.bin factory-AAEO.bin
++  IMAGE/loader.bin := append-loader-okli $(1) | pad-to 64k | lzma | \
++	uImage lzma
++  IMAGE/factory-AAEO.bin := $$(IMAGE/sysupgrade.bin) | pad-to 8192k | \
++	zyxel-tar-bz2 zyxel_nwa1123-ni vmlinux_mi124_f1e mi124_f1e-jffs2 | \
++	append-md5sum-bin
++endef
++TARGET_DEVICES += zyxel_nwa1123-ni
++
+ define Device/zyxel_nbg6616
+   SOC := qca9557
+   DEVICE_VENDOR := ZyXEL
+-- 
+2.25.1
+
diff --git a/targets/ath79-generic b/targets/ath79-generic
index 246baabc..5724b98a 100644
--- a/targets/ath79-generic
+++ b/targets/ath79-generic
@@ -94,3 +94,9 @@ device('tp-link-cpe220-v3', 'tplink_cpe220-v3')
 device('tp-link-tl-wdr3600-v1', 'tplink_tl-wdr3600-v1')
 
 device('tp-link-tl-wdr4300-v1', 'tplink_tl-wdr4300-v1')
+
+-- ZyXEL
+
+device('zyxel-nwa1121-ni', 'zyxel_nwa1121-ni')
+
+device('zyxel-nwa1123-ni', 'zyxel_nwa1123-ni')
-- 
2.25.1

